crates/rethnet_evm/src/debug_trace.rs

```diff
@@ -0,0 +1,314 @@
+use crate::blockchain::SyncBlockchain;
+use crate::evm::build_evm;
+use crate::state::SyncState;
+use crate::TransactionError;
+use rethnet_eth::transaction::TransactionRequest;
+use rethnet_eth::B256;
+use revm::inspectors::GasInspector;
+use revm::interpreter::{
+    opcode, CallInputs, CreateInputs, Gas, InstructionResult, Interpreter, Stack,
+};
+use revm::primitives::{hex, B160, U256};
+use revm::primitives::{BlockEnv, Bytes, CfgEnv, ExecutionResult, ResultAndState, SpecId};
+use revm::{EVMData, Inspector};
+use std::fmt::Debug;
+
+/// Get trace output for `debug_traceTransaction`
+#[cfg_attr(feature = "tracing", tracing::instrument)]
+pub fn debug_trace_transaction<BlockchainErrorT, StateErrorT>(
+    blockchain: &dyn SyncBlockchain<BlockchainErrorT, StateErrorT>,
+    state: &mut dyn SyncState<StateErrorT>,
+    cfg: CfgEnv,
+    block_env: BlockEnv,
+    transactions: Vec<TransactionRequest>,
+    transaction_hash: B256,
+) -> Result<DebugTraceResult, DebugTraceError<BlockchainErrorT, StateErrorT>>
+where
+    BlockchainErrorT: Debug + Send + 'static,
+    StateErrorT: Debug + Send + 'static,
+{
+    if cfg.spec_id < SpecId::SPURIOUS_DRAGON {
+        // Matching Hardhat Network behaviour: https://github.com/NomicFoundation/hardhat/blob/af7e4ce6a18601ec9cd6d4aa335fa7e24450e638/packages/hardhat-core/src/internal/hardhat-network/provider/vm/ethereumjs.ts#L427
+        return Err(DebugTraceError::InvalidSpecId {
+            spec_id: cfg.spec_id,
+            message: "`debug_traceTransaction` is not supported prior to Spurious Dragon"
+                .to_string(),
+        });
+    }
+
+    if cfg.spec_id > SpecId::MERGE && block_env.prevrandao.is_none() {
+        return Err(TransactionError::MissingPrevrandao.into());
+    }
+
+    // Clone the state so that we can apply throw-away modifications on it
+    // TODO figure out why this doesn't work
+    // let mut state = state.clone();
```

Comment:
You need to pass in the `Box` itself, to be able to clone, which means the clone needs to happen external to this function

Reply:
Will need the changes from https://github.com/NomicFoundation/hardhat/pull/4254